File: platforms/ios-consumer/SparrowInvest/Models/Fund.swift
import Foundation

struct Fund: Codable, Identifiable {
    let id: String
    let schemeCode: Int
    let schemeName: String
    let category: String
    let assetClass: String
    var nav: Double
    var navDate: Date?
    var returns: FundReturns?
    var aum: Double?
    var expenseRatio: Double?
    var riskRating: Int? // 1-5 scale
    var minSIP: Double
    var minLumpSum: Double
    var fundManager: String?
    var fundHouse: String?
    // Risk data from Kuvera
    var isin: String?
    var crisilRating: String?   // Original CRISIL text (e.g., "Moderately High Risk")
    var volatility: Double?     // Percentage

    var shortName: String {
        // Extract a shorter version of the fund name
        let words = schemeName.components(separatedBy: " ")
        if words.count > 4 {
            return words.prefix(4).joined(separator: " ")
        }
        return schemeName
    }

    var initials: String {
        let words = schemeName.components(separatedBy: " ")
        return words.prefix(2).compactMap { $0.first }.map(String.init).joined()
    }
}

struct FundReturns: Codable {
    var oneMonth: Double?
    var threeMonth: Double?
    var sixMonth: Double?
    var oneYear: Double?
    var threeYear: Double?
    var fiveYear: Double?

    enum CodingKeys: String, CodingKey {
        case oneMonth = "1m"
        case threeMonth = "3m"
        case sixMonth = "6m"
        case oneYear = "1y"
        case threeYear = "3y"
        case fiveYear = "5y"
    }
}

struct FundRecommendation: Codable, Identifiable {
    let id: String
    let fund: Fund
    let score: Double
    let reasons: [String]
    let suggestedAllocation: Double

    var topReason: String {
        reasons.first ?? "Matches your profile"
    }
}

// MARK: - Fund Categories
enum FundCategory: String, CaseIterable {
    case equity = "Equity"
    case debt = "Debt"
    case hybrid = "Hybrid"
    case elss = "ELSS"
    case index = "Index"
    case gold = "Gold"

    var description: String {
        switch self {
        case .equity: return "High growth potential, higher risk"
        case .debt: return "Stable returns, lower risk"
        case .hybrid: return "Balanced mix of equity and debt"
        case .elss: return "Tax saving with 3-year lock-in"
        case .index: return "Track market indices passively"
        case .gold: return "Hedge against inflation"
        }
    }

    var icon: String {
        switch self {
        case .equity: return "chart.line.uptrend.xyaxis"
        case .debt: return "shield.fill"
        case .hybrid: return "circle.lefthalf.filled"
        case .elss: return "indianrupeesign.square.fill"
        case .index: return "chart.bar.fill"
        case .gold: return "dollarsign.circle.fill"
        }
    }

    var color: String {
        switch self {
        case .equity: return "#2563EB"
        case .debt: return "#10B981"
        case .hybrid: return "#F59E0B"
        case .elss: return "#8B5CF6"
        case .index: return "#14B8A6"
        case .gold: return "#EAB308"
        }
    }
}

File: platforms/android-consumer/app/src/main/java/com/sparrowinvest/app/data/model/Fund.kt
package com.sparrowinvest.app.data.model

import androidx.compose.ui.graphics.Color
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Fund(
    val id: String = "",
    @SerialName("scheme_code")
    val schemeCode: Int,
    @SerialName("scheme_name")
    val schemeName: String,
    val category: String = "",
    @SerialName("asset_class")
    val assetClass: String = "equity",
    val nav: Double = 0.0,
    @SerialName("nav_date")
    val navDate: String? = null,
    val returns: FundReturns? = null,
    val aum: Double? = null,
    @SerialName("expense_ratio")
    val expenseRatio: Double? = null,
    @SerialName("risk_rating")
    val riskRating: Int? = null, // 1-5 scale
    @SerialName("min_sip")
    val minSip: Double = 500.0,
    @SerialName("min_lumpsum")
    val minLumpsum: Double = 5000.0,
    @SerialName("fund_manager")
    val fundManager: String? = null,
    @SerialName("fund_house")
    val fundHouse: String? = null,
    // Risk data from Kuvera
    val isin: String? = null,
    @SerialName("crisil_rating")
    val crisilRating: String? = null,  // Original CRISIL text (e.g., "Moderately High Risk")
    val volatility: Double? = null     // Percentage
)

@Serializable
data class FundDetail(
    val id: String = "",
    @SerialName("scheme_code")
    val schemeCode: Int,
    @SerialName("scheme_name")
    val schemeName: String,
    val category: String = "",
    @SerialName("asset_class")
    val assetClass: String = "equity",
    val nav: Double = 0.0,
    @SerialName("nav_date")
    val navDate: String? = null,
    val returns: FundReturns? = null,
    val aum: Double? = null,
    @SerialName("expense_ratio")
    val expenseRatio: Double? = null,
    @SerialName("risk_rating")
    val riskRating: Int? = null,
    @SerialName("min_sip")
    val minSip: Double = 500.0,
    @SerialName("min_lumpsum")
    val minLumpsum: Double = 5000.0,
    @SerialName("fund_manager")
    val fundManager: String? = null,
    @SerialName("fund_house")
    val fundHouse: String? = null,
    // Risk data from Kuvera
    val isin: String? = null,
    @SerialName("crisil_rating")
    val crisilRating: String? = null,
    val volatility: Double? = null,
    val description: String? = null,
    @SerialName("inception_date")
    val inceptionDate: String? = null,
    val benchmark: String? = null,
    @SerialName("nav_history")
    val navHistory: List<NavPoint>? = null,
    @SerialName("sector_allocation")
    val sectorAllocation: Map<String, Double>? = null,
    @SerialName("top_holdings")
    val topHoldings: List<HoldingInfo>? = null
)

@Serializable
data class FundReturns(
    @SerialName("1m")
    val oneMonth: Double? = null,
    @SerialName("3m")
    val threeMonth: Double? = null,
    @SerialName("6m")
    val sixMonth: Double? = null,
    @SerialName("1y")
    val oneYear: Double? = null,
    @SerialName("3y")
    val threeYear: Double? = null,
    @SerialName("5y")
    val fiveYear: Double? = null
)

@Serializable
data class NavPoint(
    val date: String,
    val nav: Double
)

@Serializable
data class HoldingInfo(
    val name: String,
    val percentage: Double,
    val sector: String? = null
)

enum class FundCategory(val displayName: String, val color: Long) {
    EQUITY("Equity", 0xFF2563EB),
    DEBT("Debt", 0xFF10B981),
    HYBRID("Hybrid", 0xFFF59E0B),
    ELSS("ELSS", 0xFF8B5CF6),
    INDEX("Index", 0xFF14B8A6),
    GOLD("Gold", 0xFFEAB308);

    companion object {
        fun fromString(value: String): FundCategory {
            return entries.find {
                it.name.equals(value, ignoreCase = true) ||
                it.displayName.equals(value, ignoreCase = true)
            } ?: EQUITY
        }
    }
}

enum class AssetClass(val displayName: String, val color: Long) {
    EQUITY("Equity", 0xFF3B82F6),
    DEBT("Debt", 0xFF10B981),
    HYBRID("Hybrid", 0xFF8B5CF6),
    GOLD("Gold", 0xFFF59E0B),
    OTHER("Other", 0xFF64748B);

    companion object {
        fun fromString(value: String): AssetClass {
            return entries.find {
                it.name.equals(value, ignoreCase = true) ||
                it.displayName.equals(value, ignoreCase = true)
            } ?: OTHER
        }
    }
}
